ARC: fault: Fix uninitialized memory access

franc0is opened the pull request:
Found this via static analysis. When CONFIG_MULTITHREADING is not set, `guard_end` is left uninitialized and is subsequently used in a comparison.

ruuddw commented:
Thanks, this indeed looks like potentially causing random results for the check if an exception was due to stack checking or not when multithreading is disabled.
@evgeniy-paltsev, any way to determine the stack bounds in that case?

franc0is commented:
the Cortex-M port does the following:
```
	if (IS_MPU_GUARD_VIOLATION((uint32_t)z_main_stack,
			guard_len,
			fault_addr, psp)) {
		/* Thread stack corruption */
		return (uint32_t)K_THREAD_STACK_BUFFER(z_main_stack);
	}
```

Which I think would translate to:
```
guard_start = z_main_stack - Z_ARC_STACK_GUARD_SIZE;
guard_end = z_main_stack;
```

I'm happy to update the PR with the above, but would appreciate someone confirming that this sounds right given how little I know about the ARC code.

ruuddw commented:
That sounds about right, yes. In arch/arc/core/thread.c the main_stack is set up similarly but with some macros for alignment handling:
```
	void *main_stack = (K_THREAD_STACK_BUFFER(z_main_stack) +
			    K_THREAD_STACK_SIZEOF(z_main_stack));
```
Though not sure if anyone is using the stack guards without multi-threading, I believe it can be configured and it would make sense, e.g. for a bootloader.

franc0is commented:
FWIW I made the suggested change

evgeniy-paltsev commented:
Hi @franc0is,

thank you for the change. 

Currently you have different email in commit author (`Francois Baldassari <francois@fbo-mba-m3.local>`) and in Signed-off-by (`François Baldassari <francois@memfault.com>`) which is not allowed by Zephyr projects - so the CI is blocking PR merge. Could you please fix that? Thanks!

There is a bit more details about it: https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

evgeniy-paltsev commented on arch/arc/core/fault.c:
```diff
@@ -90,6 +94,9 @@ static bool z_check_thread_stack_fail(const uint32_t fault_addr, uint32_t sp)
 		guard_start = guard_end - Z_ARC_STACK_GUARD_SIZE;
 	}
+#else /* CONFIG_MULTITHREADING */
+	guard_start = (uint32_t)z_main_stack - Z_ARC_STACK_GUARD_SIZE;
+	guard_end = (uint32_t)z_main_stack;
```
Hmm, from the first look it should be something like this:

```suggestion
	guard_end = (uint32_t)z_main_stack + Z_ARC_STACK_GUARD_SIZE;
	guard_start = (uint32_t)z_main_stack;
```

franc0is replied:
@evgeniy-paltsev thank you, I have no idea what the stack layout is or where the stack guards are on ARC. I will therefore defer to you, but I will note that the other stacks assume the stack guard to be right before the stack if I read the code correctly:

```
/* Special case: handling a syscall on privilege stack.
 * There is guard memory reserved immediately before
 * it.
 */
guard_end = thread->arch.priv_stack_start;
guard_start = guard_end - Z_ARC_STACK_GUARD_SIZE;
```
and
```
/* Supervisor thread */
guard_end = thread->stack_info.start;
guard_start = guard_end - Z_ARC_STACK_GUARD_SIZE;
```

franc0is replied:
this would suggest:
```
guard_end = (uint32_t)z_main_stack;
guard_start = guard_end - Z_ARC_STACK_GUARD_SIZE;
```

franc0is commented:
Thank you for the note about the commit author (new laptop...). I've fixed it.

ruuddw commented:
I think the original proposed solution was correct (start = end - size). Confusion is about stack top and bottom vs start and end. In the check logic here, start should be the smaller address, and end the larger one.
My point earlier was that I think some alignment and optional reserved space should be taken into account:

```
 * It is NOT legal to take the sizeof(sym) and pass that to the stackSize
 * parameter of k_thread_create(), it may not be the same as the
 * 'size' parameter. Use K_THREAD_STACK_SIZEOF() instead.
 *
 * Some arches may round the size of the usable stack region up to satisfy
 * alignment constraints. K_THREAD_STACK_SIZEOF() will return the aligned
 * size.
```

franc0is commented:
Thank you for the suggestion @ruuddw. I've made the suggested change. Note that this PR is getting a bit more complicated as a result.

ruuddw commented:
Hi @franc0is, I finally had time to look a bit deeper into this, and without multithreading the MPU stack guards are not configured/supported.
So the `z_check_thread_stack_fail()` function can simply return `false` in that case, as there is no stack guard region defined for the main function.

franc0is commented:
Hah! So back to the original. I sure learned a lot in this process, thank you for your all your help!

franc0is committed:
ARC: fault: Fix uninitialized memory access

Found via static analysis. In fault path when checking for stack
overflows, if CONFIG_MULTITHREADING is not set, `guard_end` is left
uninitialized and is subsequently used in a comparison.

The solution is to simply return `false` in this configuration as stack
guards are not configured in the first place.

Signed-off-by: François Baldassari <francois@memfault.com>

evgeniy-paltsev commented:
Hi @ruuddw,

> without multithreading the MPU stack guards are not configured/supported.

Are you sure about it? As for now we define main stack the same way when we don't have multi threading: https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/init.c#L53

We access the stack area via K_THREAD_STACK_BUFFER & K_THREAD_STACK_SIZEOF - so we take MPU guard area into account (if MPU_STACK_GUARD is enabled): https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/thread.c#L289-L290

It's done deep inside these macros with adding/removing of `K_KERNEL_STACK_RESERVED`: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/kernel/thread_stack.h#L288

And `K_KERNEL_STACK_RESERVED` [is defined to ARCH_KERNEL_STACK_RESERVED](https://github.com/zephyrproject-rtos/zephyr/blob/c2c95e18ed01b642b46b118f610b29f10a2227e3/include/zephyr/kernel/thread_stack.h#L92)

And `ARCH_KERNEL_STACK_RESERVED` [is defined to Z_ARC_STACK_GUARD_SIZE](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/arc/arch.h#L145)


I don't see that MPU_STACK_GUARD depend from multi-threading. So I'd say it's supported even in single thread mode.

ruuddw commented:
Yes, the guard area is defined, but the MPU doesn't get programmed to enforce it as that is done when a thread is initialized or swapped. At least, that is the conclusion I came to.

ruuddw reviewed (approved):
Thanks!

evgeniy-paltsev commented:
> Yes, the guard area is defined, but the MPU doesn't get programmed to enforce it as that is done when a thread is initialized or swapped. At least, that is the conclusion I came to.

Yep, can confirm it.
